File: Python/Data_Structure_and_Algorithm/Search/Interpolation_search.py
def interpolation_search(arr, target):
    low, high = 0, len(arr) - 1
    
    # Continue searching while the target is within the range of the array
    while low <= high and arr[low] <= target <= arr[high]:
        # Calculate the position using interpolation formula
        position = low + ((target - arr[low]) * (high - low)) // (arr[high] - arr[low])
        
        if arr[position] == target:
            return position  # If the target is found at the calculated position, return the index
        
        elif arr[position] < target:
            low = position + 1  # If the target is greater than the value at the calculated position, search in the right half
        
        else:
            high = position - 1  # If the target is less than the value at the calculated position, search in the left half
    
    return -1  # If the target is not found, return -1

File: C:C++/Data_Structure_and_Algorithm/Search/interpolation_search.cc
#include <iostream>
using namespace std;

// Function to perform interpolation search
int interpolationSearch(int arr[], int n, int target) {
    int low = 0;
    int high = n - 1;

    while (low <= high && target >= arr[low] && target <= arr[high]) {
        // Calculate the probable position of the target
        int pos = low + ((double)(high - low) / (arr[high] - arr[low])) * (target - arr[low]);

        // If target is found at pos
        if (arr[pos] == target) {
            return pos;
        }

        // If target is less, search in the left subarray
        if (arr[pos] > target) {
            high = pos - 1;
        }
        // If target is more, search in the right subarray
        else {
            low = pos + 1;
        }
    }

    // If target is not present in array
    return -1;
}

int main() {
    int arr[] = {10, 20, 30, 40, 50, 60, 70, 80, 90, 100};
    int n = sizeof(arr) / sizeof(arr[0]);
    int target = 50;
    int result = interpolationSearch(arr, n, target);
    if (result != -1) {
        cout << "Element found at index " << result << endl;
    } else {
        cout << "Element not found" << endl;
    }
    return 0;
}